[CDAP-7193] Give program lifecycle methods control over transactions

anew opened the pull request:
The design is documented in https://issues.cask.co/browse/CDAP-7193.
It is easiest to review this commit by commit.
Every commit message describes what that commit is doing. 
Build is running here: http://builds.cask.co/browse/CDAP-DUT4923

anew commented:
Build passed http://builds.cask.co/browse/CDAP-DUT4923-15

anew commented:
This became large due to all the unit test fixes in the last commit. The change there is mostly to execute dataset operations in transactions.

awholegunch commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/TransactionPolicy.java:
```diff
@@ -0,0 +1,51 @@
+ * By default, program life cycle methods ({@link ProgramLifecycle#initialize(RuntimeContext) initialize},
+ * {@link ProgramLifecycle#destroy() destroy}), are run inside a transaction. Some methods do not require a
+ * transaction, or they have a need to control their transactions themselves. Such a method can be annotated
+ * as follows, for example in a worker:
```
```
* transaction, or they need to control their transactions themselves. Such a method can be annotated,
* for example in a worker, with:
```

anew replied:
ok

awholegunch commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/TransactionPolicy.java:
```diff
@@ -0,0 +1,51 @@
+ * }
+ * </code></pre>
+ *
+ * This is especially useful if the method may run longer than the transaction timeout, and it would therefore
```
and it would then

anew replied:
sure

awholegunch commented on cdap-api/src/main/java/co/cask/cdap/api/customaction/AbstractCustomAction.java:
```diff
@@ -15,6 +15,9 @@
 package co.cask.cdap.api.customaction;
 
+import co.cask.cdap.api.annotation.TransactionControl;
+import co.cask.cdap.api.annotation.TransactionPolicy;
```
It looks like these two classes are always being imported together. Can they be combined into one?

anew replied:
Not really. One is an annotation and should not contain other classes. You could, but it is not common practice.

awholegunch commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/flow/FlowletRuntimeService.java:
```diff
@@ -105,48 +109,66 @@ void resume() {
+      }
+      LOG.info("Flowlet initialized: " + flowletContext);
+    } catch (Throwable cause) {
       LOG.error("Flowlet throws exception during flowlet initialize: " + flowletContext, cause);
```
Should this be:

```
"Flowlet threw exception during flowlet initialization: "
```

anew replied:
That was actually not a part of this change but existing code. But I can change it, of course.

awholegunch commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceLifecycleContext.java:
```diff
@@ -314,6 +316,16 @@ public SecureStoreData getSecureData(String namespace, String name) throws Excep
+
+  @Override
+  public void execute(int timeoutInSeconds, TxRunnable runnable) throws TransactionFailureException {
+    throw new TransactionFailureException("Attempt to start a transaction within a MapReduce transaction");
```
Should this be:

```
"Attempted to start a transaction within a MapReduce transaction"
```

anew replied:
sure.

awholegunch commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java:
```diff
@@ -526,6 +539,18 @@ public Void call() throws Exception {
+  private void doInitialize() throws Exception {
+    context.setState(new ProgramState(ProgramStatus.INITIALIZING, null));
+    if (mapReduce instanceof ProgramLifecycle) {
+      //noinspection unchecked
```
Unclear what this comment is saying.

anew replied:
it is a directive to suppress a warning.

awholegunch commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/SingleThreadDatasetCache.java:
```diff
@@ -255,7 +255,11 @@ private void discardSafely(Object dataset) {
-  public TransactionContext newTransactionContext() {
+  public TransactionContext newTransactionContext() throws TransactionFailureException {
+    if (txContext != null && txContext.getCurrentTransaction() != null) {
+      throw new TransactionFailureException("Attempt to start a transaction within active transaction " +
```
```
"Attempted to start a transaction within active transaction "
```

anew replied:
sure

awholegunch commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/SingleThreadDatasetCache.java:
```diff
@@ -403,11 +407,10 @@ public void start(int timeout) throws TransactionFailureException {
-        cleanup();
+    private void newTxContext() throws TransactionFailureException {
+      if (getCurrentTransaction() != null) {
+        throw new TransactionFailureException("Attempt to start a transaction within active transaction " +
```
ibid.

anew replied:
ok

awholegunch commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/BufferingTable.java:
```diff
@@ -363,9 +368,16 @@ public boolean rollbackTx() throws Exception {
 
+  protected void ensureTransactionIsStarted() {
+    if (tx == null) {
+      throw new DataSetException("Attempt to perform a data operation without a transaction");
```
```
"Attempted to ...
```

anew replied:
ok

awholegunch commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/Transactions.java:
```diff
@@ -344,4 +348,59 @@ public static RuntimeException propagate(TransactionFailureException e) {
+   * @param baseClass upper bound for the super classes to be inspected
+   * @param program the program
+   * @param methodName the name of the method
+   * @param params the paramters of the method
```
the parameters

anew replied:
done

awholegunch commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/Transactions.java:
```diff
@@ -344,4 +348,59 @@ public static RuntimeException propagate(TransactionFailureException e) {
+   * @param params the paramters of the method
+   *
+   * @return the transaction control annotated for the method of the program class or its nearest superclass that
+   *         is annotated; or defaultValue if no annotation is present in any of the superclasses.
```
no trailing period

anew replied:
ok

anew commented:
I have removed the last two commits (making sure that no data operations are performed outside of a transaction) because they doubled the size of this PR and could easily be moved to a separate branch.

anew commented:
@poornachandra PR is now only half the size. Please review. Thanks

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/customaction/AbstractCustomAction.java:
```diff
@@ -58,6 +61,7 @@ protected void setProperties(Map<String, String> properties) {
   }
 
   @Override
+  @TransactionPolicy(TransactionControl.IMPLICIT)
```
I see the addition of this annotation across all abstract program classes. Is it necessary or is it just a way to document what's the default transaction policy is?

anew replied:
Yes, I added that to make it clear what the default is for each program type. The program runners do use the same as default.

chtyim replied:
Shall the annotation on the top-level interface then and have the `TransactionPolicy` annotation annotated with `@Inherited` as well? That makes the contract clear starting from the top-level.

https://docs.oracle.com/javase/7/docs/api/java/lang/annotation/Inherited.html

chtyim replied:
E.g. we can have it at the `ProgramLifecycle` level with `implicit` and only `Worker` and `CustomActionContext` overrides it to annotate with `explicit`.

anew replied:
we can do that. But the initialize() without arguments (in MR, Spark, CustomAction) still needs the annotation

chtyim replied:
Yes. That's what I meant.

anew replied:
Done, except that CustomAction is also implicit.

poornachandra commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/DynamicTransactionExecutor.java:
```diff
@@ -118,7 +118,16 @@ public Void apply(Void input) throws Exception {
+    TransactionContext txContext;
+    try {
+      txContext = txContextSupplier.get();
+    } catch (RuntimeException e) {
```
We can just catch `TransactionFailureException` over there, right?

anew replied:
supplier.get() cannot throw checked exceptions. Therefore, the context supplier must wrap a possible TransactionFailureException into a runtime exception.

this can happen, for example, if the supplier has an active transaction, and this would represent a nested transaction.

I added this as a comment to the code

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/customaction/AbstractCustomAction.java:
```diff
@@ -67,11 +71,13 @@ public final void initialize(CustomActionContext context) throws Exception {
    * Classes derived from {@link AbstractCustomAction} can override this method to initialize the {@link CustomAction}.
    * @throws Exception if there is any error in initializing the custom action
    */
+  @TransactionPolicy(TransactionControl.IMPLICIT)
```
So, this method is called from `initialize(CustomActionContext context)` which is a final method with `implicit` policy. What happen if the user implements this method but change the transaction policy to `explicit`? What will happen? The platform only aware of the `initialize(context)` method, but not this method, which means I believe an `implicit` transaction will always be created before the call.

anew replied:
That is why we have special handling in the CustomActionExecutor: it checks whether the action extends AbstractCustomAction, and if so, it determines the transaction policy of initialize(). 

That means, for certain program types (MR, Spark, CustomAction), we have special handling in the platform, which is aware that the abstract program implement a final initialize(context) and requires initialize() from the subclass instead.

This is covered by test cases.

anew replied:
I also added comments in the code

poornachandra commented on cdap-unit-test/src/test/java/co/cask/cdap/test/app/AppWithCustomTx.java:
```diff
@@ -0,0 +1,316 @@
+      LOG.error("Nested transaction should not have succeeded for {}:{}", row, key);
+    } catch (TransactionFailureException e) {
+      // expected: starting nested transaction should fail
+      LOG.info("Nested transaction failed as expected for {}:{}", row, key);
```
It would be good to record a failure in the dataset using the existing transaction.

anew replied:
very good point. added that.

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java:
```diff
@@ -496,28 +498,39 @@ private Location createTempLocationDirectory() throws IOException {
   @SuppressWarnings("unchecked")
   private void beforeSubmit(final Job job) throws TransactionFailureException {
+
+    final TransactionControl txControl = mapReduce instanceof AbstractMapReduce
```
Curious as to why similar check is not done for `destroy` method below?

anew replied:
AbstractMapReduce implements initialize(context) as final and requires that subclasses implement initialize(), without arguments. For destroy, AbstractMapReduce simply defines a no-op method. So it does not matter whether the program implements AbstractMapReduce.

poornachandra reviewed (commented):
@anew I did a pass over the PR and have a couple of comments, will do another pass later.

anew commented:
@poornachandra @chtyim I have addressed your comments. I also rebased on latest develop to fix a build break.

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/Transactions.java:
```diff
@@ -344,4 +348,59 @@ public static RuntimeException propagate(TransactionFailureException e) {
+  private static TransactionControl getTransactionControl(Class<?> cls, String methodName, Class<?>[] params) {
+    try {
+      Method method = cls.getDeclaredMethod(methodName, params);
+      Annotation annotation = method.getAnnotation(TransactionPolicy.class);
```
You can write it as `TransactionPolicy txPolicy = method.getAnnotation(TransactionPolicy.class);` to avoid casting later.

anew replied:
done

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/Transactions.java:
```diff
@@ -344,4 +348,59 @@ public static RuntimeException propagate(TransactionFailureException e) {
+        break; // reached bounding superclass
+      }
+      cls = cls.getSuperclass();
+      if (cls == null) {
```
We can stop at `Object.class` instead of the parent of `Object.class` (i.e. null), right?

anew replied:
right

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java:
```diff
@@ -496,28 +498,42 @@ private Location createTempLocationDirectory() throws IOException {
+    final TransactionControl txControl = mapReduce instanceof AbstractMapReduce
+      ? Transactions.getTransactionControl(TransactionControl.IMPLICIT, AbstractMapReduce.class,
+                                           mapReduce, "initialize")
+      : mapReduce instanceof ProgramLifecycle
```
Actually we don't need this condition, right? Can just do:

``` java
txControl = mapReduce instanceof AbstractMapReduce
  ? Transactions.getTransactionControl(TransactionControl.IMPLICIT, AbstractMapReduce.class, mapReduce, "initialize")
  : Transactions.getTransactionControl(TransactionControl.IMPLICIT, MapReduce.class, mapReduce, "initialize", MapReduceContext.class);
```

Because if the mapreduce program is not extended from `AbstractMapReduce` and doesn't implement `ProgramLifecycle`, the second call to `getTransactionControl` (the one after `:`) will default it to `IMPLICIT` anyway.

chtyim replied:
However, with the extra condition, it does save some runtime check for the annotation. It's up to you.

anew replied:
I also think it is more explicit when reading this code.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java:
```diff
@@ -496,28 +498,42 @@ private Location createTempLocationDirectory() throws IOException {
+      try {
+        doInitialize();
+      } catch (Exception e) {
+        throw Throwables.propagate(e);
```
Can we actually have the private `beforeSubmit` method to have `throws Exception` so that we don't have to wrap?

anew replied:
good point. It does not save us the try { } block, though, because we need the finally clause.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java:
```diff
@@ -496,28 +498,42 @@ private Location createTempLocationDirectory() throws IOException {
           setOutputsIfNeeded(job);
-
           setOutputClassesIfNeeded(job, mapperTypeToken);
           setMapOutputClassesIfNeeded(job, mapperTypeToken);
```
So these couple `setXXX` calls don't need to be inside transaction, or do they? I was a bit confused looking at this method as I expect something more like this:

``` java
TransactionControl txControl = // determine tx control like above
if (txControl == EXPLICIT) {
  doInitialize();  // assuming the setting of ClassLoader is inside the doInitialize method
} else {
  Transactions.execute(...., new Callable<Void>() {
    public Void call() throws Exception {
      doInitialize();
    }
  });
}

// Set MR job properties based on result from the initialize call
TypeToken<?> mapperTypeToken = setInputsIfNeeded(job);
......
```

anew replied:
setOutputsIfNeeded() must be inside a transaction, and so does setInputsIfNeeded()

anew replied:
we can probably refactor the setInput and setOutputIfNeeded not to require a transaction. I will open a separate Jira for that.

anew replied:
opened CDAP-7475 and CDAP-7476 for that.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java:
```diff
@@ -567,11 +595,14 @@ private ProgramState getProgramState(boolean success, String failureInfo) {
     try {
-      Transactions.execute(txContext, "destroy", new Callable<Void>() {
+      TransactionContext txContext = context.getTransactionContext();
+      boolean success = Transactions.execute(txContext, "destroy", new Callable<Boolean>() {
```
Similar to above, if the tx control is `EXPLICIT`, we are creating a tx here just to execute the `context.setState` call, which actually doesn't need to be transactional. I think it's cleaner if we just have a if-else logic for the implicit and explicit case.

anew replied:
here we do need the extra transaction, because we must call the output committers. See a few lines above.

anew replied:
created CDAP-7477 to address this.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java:
```diff
@@ -583,17 +614,25 @@ public Void call() throws Exception {
         }
       });
+      if (TransactionControl.EXPLICIT == txControl) {
+        ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(mapReduceClassLoader);
```
Better to move the setting and resetting of classloader inside the `doDestroy` method (similarly for `doInitialize`) to reduce code duplication.

anew replied:
it's duplication, but one time we execute only doDestroy() inside, and one time we execute more than that.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGenerator.java:
```diff
@@ -515,10 +515,11 @@ public void visitClassType(String name) {
      *     HttpContentConsumer contentConsumer = null;
      *     try {
+     *       txContext = getTransactionContext();
+     *       DelayedHttpServiceResponder wrappedResponder = wrapResponder(responder, txContext);
```
Seems like the comment is not align with the actual code being generated. With the change, the code generated looks like this:

``` java
TransactionContext txContext = null;
HttpContentConsumer contentConsumer = null;
DelayedHttpServiceResponder wrappedResponder = null;
try {
  txContext = getTransactionContext();
  wrappedResponder = wrapResponder(responder, txContext);
  ....
}
```

anew replied:
good catch.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java:
```diff
@@ -203,33 +205,41 @@ private Workflow initializeWorkflow() throws Exception {
+    if (TransactionControl.EXPLICIT == txControl) {
+      doInitialize(workflow);
+    }
+    basicWorkflowContext.execute(new TxRunnable() {
```
Similar to the comment in `MapReduceRuntimeService`, we don't need to run `runtimeStore.updateWorkflowToken` from a transaction.

anew replied:
fixed for destroy and initialize in this class.

chtyim commented:
Did a pass, raised couple questions.

anew commented:
Build has passed: http://builds.cask.co/browse/CDAP-DUT4923

@chtyim is it good to merge?

chtyim commented:
LGTM

Commit:
[CDAP-7193] Give program lifecycle methods a way to control the transactions:
- Add a @TransactionPolicy annotation;
- Implement it for lifecycle methods of all program types;
- Allow accessing datasets without a transaction;
- Refactor existing test case for transaction timeout to test also this

anew commented:
OK, will squash commits and merge.
